Fix `zhack label repair` checksum repair regression on nonzero TXG

buzzingwires opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context

This change is intended to fix a regression introduced by 64db435 and first noticed when reviewing [these](https://github.com/openzfs/zfs/issues/4187#issuecomment-3215969585) error messages.

The aforementioned commit aimed to maintain the `zhack label repair` behavior of d04b5c9 by default, but performed checks and unnecessarily accessed data associated with the new `-u` (undetach) instruction when doing so.

This caused the old checksum repair behavior (default or `-c`) to fail when an uberblock with a nonzero TXG was encountered, when this behavior was only intended for `-u`.

The issue was not caught by old tests because there are so few transactions made on a testing zpool that some uberblocks are never assigned a TXG.

### Description

* The check for an uberblock's TXG being zero is now only performed for the `-u` option, which assumes a clean detach.

* The ashift value is only accessed when needed by the `-u` option.

* Determining whether to do byteswap for checksum calculation fails if the magic number doesn't match the expected for either endianness, rather than only if the magic number is zero.

* Code is refactored to better separate it into discrete tasks and track return values.

* In a separate commit, `typeset`s are added to local variable declarations in test scripts to enforce their scope.

### How Has This Been Tested?

Case one (`zhack_label_repair_001.ksh`) of the relevant tests has been revised to do the following ~~128~~ 256 times: `touch` the test data stored in the pool, then call `zpool sync` on the pool. This ensures there are enough updates for every uberblock to be used, so that the checksum repair will encounter a nonzero TXG. Documentation for these tests is updated accordingly.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [X] Code cleanup (non-breaking change which makes code smaller or more readable)
- [X] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [X] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [X] I have run the ZFS Test Suite with this change applied.
- [X] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

buzzingwires committed:
Refactor `zhack label repair` and fix `-c` regression on nonzero TXG

This commit fixes a likely regression introduced by 64db435 where the
checksum repair functionality (`-c` or default behavior) will perform
checks and access data associated with the newer undetach (`-u`)
functionality, resulting in a failure when an uberblock's TXG is not 0
as required by `-u` but not `-c`

Additionally, code is refactored for better separation of tasks.

Signed-off-by: buzzingwires <buzzingwires@outlook.com>

buzzingwires commented:
Thank you for the review. Question about the shell dialect: Should I generally assume `ksh93` will be used to run tests for OpenZFS? This is very much my oversight (and lack of familiarity with ksh), but I've realized `typeset` should probably be used to give variables local scope. Probably benign *at the moment*, but not great from a QA standpoint.

behlendorf commented:
Yes, for historical reasons the test suite is almost all in ksh93.  Using `typeset` is probably technically correct here, but given how small the test scripts are in general it's rarely been a problem.  I don't think it's critical to add the missing `typeset`s, but if you'd like to add them that'd be great.

buzzingwires commented:
Totally understand. Just wanted to know what I was working with. I certainly will, then. Would you prefer on this PR?

buzzingwires committed:
Add `typeset`s to `zhack label repair` test scripts

As a quality assurance measure, `typeset` is added to local variable
declarations to actually enforce their intended scope.

Signed-off-by: buzzingwires <buzzingwires@outlook.com>

behlendorf commented:
Sure, including it in this PR would be fine.  Please just keep that change separate as its own commit.

SeDemal commented:
Hi, 
I struggled with the same error that prompted this PR (UB TXG expecting 0), running in circles trying to fix my label's checksum for days on end even though I had the right command already.
I stumbled on this pr by chance and had to manually compile it on my machine so I could use the corrected zhack label repair tool.
@buzzingwires You saved my whole server, thank you.

Now, I am just a bit perplexed as to why this isn't merged, and even closed? I see the closing is reffering a commit so maybe it's already implemented somewhere in another PR and it hasn't made it's way to my OS (Truenas scale).

gmelikov commented:
@SeDemal it's merged in master branch, but it's not in any release yet. So it'll be in next major release (2.5?).
